### PR TITLE
Only show and use filters from enabled plugins

### PR DESCRIPTION
**Pull request by nick29581**

### CODE REVIEW

**Comment by nick29581** [comment]
@erikrose ping - any update here?

**Comment by erikrose** [comment]
Sorry, trying to get the ES branch polished for staging. I'll have a look tomorrow!

**Comment by nick29581** [comment]
@erikrose ping - changes made, see last commit

**Comment by erikrose** [comment]
Having a look now.

File: dxr/query.py [hunk 1]
```diff
@@ -309,17 +320,51 @@ def generic_visit(self, node, visited_children):
         """
         return visited_children or node
 
+def filters_per_plugin(enabled_plugins):
```
Comment (erikrose) [inline dxr/query.py]: It's not clear from the docstrings what this and the next 2 functions do. Trying to figure it out....

File: dxr/query.py [hunk 2]
```diff
@@ -309,17 +320,51 @@ def generic_visit(self, node, visited_children):
...
+    """Return a list of filters for each plugin in enabled_plugins"""
+    # A list of plugins for all enabled plugins
+    plugins = [p for (n, p) in all_plugins().iteritems() if n in enabled_plugins]
+    # A list of filters (with a non-null description) for each enabled plugin 
+    filters = [[f for f in p.filters if f.description] for p in plugins]
```
Comment (erikrose) [inline dxr/query.py]: Testing for "if f.description" here breaks plain-text searching.

File: dxr/query.py [hunk 3]
```diff
@@ -179,52 +172,70 @@ def direct_result(self):
...
-        # regexp, function, etc. No filter is a prefix of a later one. This
-        # avoids premature matches.
-        '|'.join(sorted((re.escape(filter_name) for
-                                filter_name, filters in
-                                FILTERS_NAMED.iteritems() if
-                                filters[0].description),
-                        key=len,
-                        reverse=True)) + ur'''"
+query_grammar_cache = {}
+@cached(query_grammar_cache)
```
Comment (erikrose) [inline dxr/query.py]: Much better. I did another round of simplification so you no longer have to pass in the cache from outside as well.

**Comment by nick29581** [comment]
updated with (hopefully) better docstrings and moved the valid description check to fix the plain text search bug

**Comment by erikrose** [comment]
What part of "merged" don't you understand, GitHub?

**Pull request closed without merging**